wfe/web-front-end.go

```diff
@@ -192,29 +192,77 @@ func marshalIndent(v interface{}) ([]byte, error) {
 	return json.MarshalIndent(v, "", "  ")
 }
 
+func (wfe *WebFrontEndImpl) relativeEndpoint(request *http.Request, endpoint string) string {
+	var result string
+	proto := "http://"
+	host := request.Host
+
+	// If the request was received via TLS, use `https://` for the protocol
+	if request.TLS != nil {
+		proto = "https://"
+	}
+
+	// Allow upstream proxies  to specify the forwarded protocol. Allow this value
+	// to override our own guess.
+	if specifiedProto := request.Header.Get("X-Forwarded-Proto"); specifiedProto != "" {
+		proto = fmt.Sprintf("%s://", specifiedProto)
+	}
+
+	// Default to "localhost" when no request.Host is provided. Otherwise requests
+	// with an empty `Host` produce results like `http:///acme/new-authz`
+	if request.Host == "" {
+		host = "localhost"
+	}
+
+	if wfe.BaseURL != "" {
+		result = fmt.Sprintf("%s%s", wfe.BaseURL, endpoint)
+	} else {
+		result = fmt.Sprintf("%s%s%s", proto, host, endpoint)
+	}
+
+	return result
+}
+
+func (wfe *WebFrontEndImpl) relativeDirectory(request *http.Request, directory map[string]string) ([]byte, error) {
+	// Create an empty map sized equal to the provided directory to store the
+	// relative-ized result
+	relativeDir := make(map[string]string, len(directory))
+
+	// Copy each entry of the provided directory into the new relative map. If
+	// `wfe.BaseURL` != "", use the old behaviour and prefix each endpoint with
+	// the `BaseURL`. Otherwise, prefix each endpoint using the request protocol
+	// & host.
+	for k, v := range directory {
+		relativeDir[k] = wfe.relativeEndpoint(request, v)
+	}
+
+	directoryJSON, err := marshalIndent(relativeDir)
+
+	// This should never happen since we are just marshalling known strings
+	if err != nil {
+		return []byte{}, err
+	}
+
+	return directoryJSON, nil
+}
+
 // Handler returns an http.Handler that uses various functions for
 // various ACME-specified paths.
 func (wfe *WebFrontEndImpl) Handler() (http.Handler, error) {
-	wfe.NewReg = wfe.BaseURL + NewRegPath
-	wfe.RegBase = wfe.BaseURL + RegPath
-	wfe.NewAuthz = wfe.BaseURL + NewAuthzPath
-	wfe.AuthzBase = wfe.BaseURL + AuthzPath
-	wfe.ChallengeBase = wfe.BaseURL + ChallengePath
-	wfe.NewCert = wfe.BaseURL + NewCertPath
-	wfe.CertBase = wfe.BaseURL + CertPath
-
-	// Only generate directory once
-	directory := map[string]string{
+	wfe.NewReg = NewRegPath
```

Comment:
Actually on second reading, I think this PR is the appropriate place to make the first change. Previously the reason for these instance fields was that BaseURL wasn't known until runtime. Now we don't need that, so we can remove those fields. Yay for deleting more code!